Purchases: Sort all user purchases by status urgency.

elliottprogrammer opened the pull request:
#### Proposed Changes

This PR sorts the purchases list at `/me/purchases` by "**Status**" (the Status column), so that purchases that are requiring some action (such as Pending activation, or Disconnected, or Plan expired, or Payment method expired, etc) will appear at the top of the list, sorted by importance/urgency.

Partially completes: https://github.com/Automattic/jetpack/issues/22913#issuecomment-1047032937
Completes task: 1164141197617539-as-1201858908716160/f

#### Implementation notes:

The Purchases are still grouped together by site (like they always have been), and then the purchases/sites that require some action (such as disconnected site, license pending activation, expiring/expired subscription, etc are sorted to the top based on importance.

The sort order looks like this:
1. Site is disconnected
2. License not activated
3. Subscription expired
4. Subscription expiring soon (or CC expires before renew date)
5. Purchases where no action is needed
6. Purchases that are managed elsewhere (by hosting partner)

Also this PR updates the status for non-activated license-based purchases from "Pending activation" to a more actionable message of, "Activate your product license key" with a _Learn more_ link pointing to a Jetpack.com page explaining how to activate product license keys. As suggested per issue https://github.com/Automattic/jetpack/issues/22913

Ideally, next steps to follow-up this PR would be to convert `/client/me/purchases-list` component to a functional component and then memoize the `getPurchasesBySite()` with `useMemo`. Probably some other optimizations also..


#### Testing Instructions

It's kind of difficult to test this PR.
The best way to test this PR is with a WordPress account with multiple connected sites, and ideally some of the sites will have 2 or more purchased products.  And also ideally you will have 1 or more purchased, unactivated product license(s).
If you don't have multiple sites with purchases, you will have to create/connect multiple Jurassic Ninja sites, and purchase a product (or more) for each of them. Then also purchase a product originating from Jetpack.com so that you will obtain a product license (But do not attach/activate it to a site).

- Do one of these:
    - Use the Calypso Live link:
        - Click the Calypso Live link below, and go to `/me/purchases`
    - Or spin up this PR by: `git fetch && git checkout add/sort-all-user-purchases && yarn start`
- Then go to `http://calypso.localhost:3000/me/purchases`
- Compare with Wordpress.com and see that all purchases needing action are sorted to the top based on importance.
- You will likely want to Disconnect a site, then reload the page to see that it gets sorted to the top. (reconnect it when done)
- You may also likely want to open your purchases/subscriptions in Store Admin(SA) and modify subscription expiration dates to force products to be expired or expiring soon (<30 days). (Remember to set them back to the way they were, if they are on real, non-testing sites you own)



#### Screenshots: (Before & After):

**BEFORE:**

![Screenshot on 2022-10-07 at 12-47-01](https://user-images.githubusercontent.com/11078128/194605701-e00617a7-8147-47e9-aa8e-360f3eda9875.png)

.
**AFTER:**

![Screenshot on 2022-10-07 at 12-47-34](https://user-images.githubusercontent.com/11078128/194605723-b1d35ac6-7cc8-4e99-8c02-5c5723ce37d8.png)



#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [X] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [X] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - 0-as-1201858908716160

elliottprogrammer committed:
Sort user purchases by status urgency.

manzoorwanijk commented on client/lib/purchases/index.ts:
```diff
@@ -41,51 +41,166 @@ import type { ReactChild } from 'react';
 const debug = debugFactory( 'calypso:purchases' );
 
-export interface SiteWithPurchases {
+interface PurchaseWithStatus extends Purchase {
+	currentPurchaseStatus: string;
+}
```
Why not simply update the `Purchase` type defined in `client/lib/purchases/types.ts` instead of creating a new type?

monsieur-z commented on client/lib/purchases/index.ts:
```diff
@@ -41,51 +41,165 @@ import type { ReactChild } from 'react';
+		( isExpiring( purchase ) &&
+			expiry < moment().add( 30, 'days' ) &&
+			! isRecentMonthlyPurchase( purchase ) ) ||
+		( isRenewing( purchase ) &&
```
Should that second part of the OR return `'paymentMethodExpired'` instead?

elliottprogrammer replied:
No, I don't believe so.
The first part of the OR is checking if the purchase is set to "manual renew" and it's expiring soon.  The second part of the OR is checking if the purchase is an "auto renew" and it's expiring soon, because when a purchase is set to 'auto renew', it won't ever expire, unless the payment method ends up expiring before the renew date.  That's what the second part of this OR is checking: it's saying, if the purchase is 'auto-renewing' and it has a renew date, and the creditCard expires before the renew date, then it's 'expiring soon'. 😉

monsieur-z replied:
I see, thanks for the explanation!

monsieur-z commented on client/lib/purchases/index.ts:
```diff
@@ -41,51 +41,165 @@ import type { ReactChild } from 'react';
+		return 'hasExpiringSoon';
+	}
+
+	return 'noPaymentActionNeeded';
```
Effectively it'd do pretty much the same, but I'm wondering if this function and the one below should return `null` by default, instead of `'noPaymentActionNeeded'`.

Let's say we introduce a new value for `purchase.currentPurchaseStatus`. If we forget to update this function to handle that new status, the returned value `noPaymentActionNeeded` would be wrong, while `null` would give us a hint that we need to update the function.

elliottprogrammer replied:
Thanks for the feedback @monsieur-z!  Hmm, if I'm understanding correctly, I don't think your suggestion will work for this scenario because the `noPaymentActionNeeded` has a _specific_ sort order position (that is not in the last position).
For example:
```
// The purchases sort order

const purchaseStatus: PurchaseStatus = {
	expired: 10,
	paymentMethodExpired: 20,
	expiringSoon: 30,
	noPaymentActionNeeded: 40,   // The 'noPaymentActionNeeded' position is Here!
	cannotManage: 100,
};
```

If we returned `null` in place of `noPaymentActionNeeded` then we would have to do something like this (which wouldn't work):

```
const purchaseStatus: PurchaseStatus = {
	expired: 10,
	paymentMethodExpired: 20,
	expiringSoon: 30,
	null: 40,   // ??? The  'noPaymentActionNeeded' position is Here! ???
	cannotManage: 100,
};
```

Let me know what you think. Or let me know if I am not understanding correctly. 😉

monsieur-z replied:
I get what you say. I guess I'm focusing on the meaning of the string `noPaymentActionNeeded`, but in truth its value doesn't really matter since it's returned by default.

monsieur-z commented:
Code changes make sense, and the PR tested well, @elliottprogrammer. I've just removed a line (see [this commit)](https://github.com/Automattic/wp-calypso/pull/67360/commits/e52a2a6bdbc01475977ab760ecd562ab5da0f865) that marked disconnected sites as connected. It might have been used for testing, and not removed?

I left some minor inline questions/comments.

ederrengifo commented:
Sorry for the late response to this PR. I only have one comment about the sorting logic for this group. Is it possible to sort them by the expiration date? I mean, the dates that are close to expiration at the top @elliottprogrammer 
<img width="585" alt="Screen Shot 2022-09-14 at 10 49 45" src="https://user-images.githubusercontent.com/10525007/190202361-d5a11d6f-d62c-463d-a117-01ba301fccc3.png">

elliottprogrammer committed:
Fix slylelint errors & warnings.

elliottprogrammer committed:
Make 'Pending activation' more actionable,
per https://github.com/Automattic/jetpack/issues/22913\#issuecomment-1244187089

ederrengifo commented:
@elliottprogrammer that makes a lot of sense. It's fine then, the core of the problem was already solve, LGTM!

ederrengifo reviewed (approved):
LGTM

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7575079).
